pysilcam/tracking/silcamtracker.py

```diff
@@ -0,0 +1,408 @@
+                        OK = True
+                        break
+                    else:
+                        OK = False
+                else:  # if there are no particles in the search box then forget it
+                    continue
+            else:
+                OK = True
+
+        if not OK:
+            continue
+
+        # if there is a particle here, use its centroid location for the vector
+        # calculation
+        cr2 = props2[int(idx - 1)].centroid  # subtract 1 from idx because of zero-indexing
+
+        # remove this particle from future calculations
+        iml2[iml2 == int(idx)] = 0
+
+        x.append(cr2[1])  # col
+        y.append(cr2[0])  # row
+
+        # and append the position of this particle from the image 1
+        y1.append(cr[0])  # row
+        x1.append(cr[1])  # col
+
+        # if we get here then we also want the particle stats
+        ecd.append(el.equivalent_diameter)
+        length.append(el.major_axis_length)
+        width.append(el.minor_axis_length)
+
+    X = [x1, x]  # horizontal vector
+    Y = [y1, y]  # vertical vector
+    return X, Y, ecd, length, width, imbw_out
+
+
+def match_last_pair(data):
+    data_1 = data[data['t2'] == data.iloc[-1].t1]
+    x_end = data_1['x2'].values
+    y_end = data_1['y2'].values
+
+    data_2 = data[data['t1'] == data.iloc[-1].t1]
+    x_start = data_2['x1'].values
+    y_start = data_2['y1'].values
+
+    c = len(x_start)
+    for i in range(c):
+        dxy = abs(x_end - x_start[i]) + abs(y_end - y_start[i])
+        ind = np.argwhere(dxy < 1e-4)
+        if len(ind) == 1:
+            data.loc[data_2.iloc[i].name, 'UPID-backward-match'] = data_1.iloc[int(ind)].name
+            data.loc[data_1.iloc[int(ind)].name, 'UPID-forward-match'] = data_2.iloc[i].name
+    return data
+
+
+def calculate_speed_df(data, PIX_SIZE):
+    X = np.array([data['x-arrival'], data['x-departure']])
+    Y = np.array([data['y-arrival'], data['y-departure']])
+
+    dt = pd.Series(np.abs(pd.to_datetime(data['t-arrival']) -
+                          pd.to_datetime(data['t-departure']))).dt.total_seconds()
+    dt = dt.values
+
+    Y_mm = Y * PIX_SIZE * 1e-3
+    X_mm = X * PIX_SIZE * 1e-3
+    dY_mm = np.diff(Y_mm, axis=0)
+    dX_mm = np.diff(X_mm, axis=0)
+    dD_mm = np.sqrt(dY_mm ** 2 + dX_mm ** 2)
+    dD_m = dD_mm / 1000
+    S_ms = dD_m / dt  # speed in m/s
+    S_cms = S_ms * 100  # speed in cm/s
+
+    Xs_mms = dX_mm / dt
+    Xs_cms = Xs_mms / 10
+
+    data['S_cms'] = S_cms[0]
+    data['X_cms'] = Xs_cms[0]
+    return data
+
+
+def extract_continuous_tracks(tracks, max_starts=None):
+    # find positions in the dataframe where there is a forward match but not a backward match
+    # these are the first occurances of a particle  ('arrivals')
+    starts = tracks[np.isnan(tracks['UPID-backward-match']) & ~np.isnan(tracks['UPID-forward-match'])].index
+
+    print('starts', len(starts))
+
+    if not max_starts is None:

```

Comment:
E714: test for object identity should be 'is not'  <!-- Automatically posted by Lintly -->